Modified API to support use without http.Server

oxtoacart opened the pull request:
For getlantern/lantern-internal#979

oxtoacart commented:
@myleshorton @joesis @uaalto This is ready for review

joesis commented on server.go:
```diff
@@ -0,0 +1,19 @@
+		if err != nil {
+			return errors.New("Unable to accept: %v", err)
+		}
+		go proxy.Handle(context.Background(), conn)
```
Maybe create the context inside `Handle()` if it's always `context.Background()`?

oxtoacart replied:
The Proxy interface is public and we want to allow other server implementations to pass their own context. This server is just a convenience, not the only entry point.

oxtoacart commented on filters/filter.go:
```diff
@@ -0,0 +1,98 @@
+	return c.apply(ctx, req, next, 0)
+}
+
+func (c Chain) apply(ctx context.Context, req *http.Request, next Next, idx int) (*http.Response, context.Context, error) {
```
Good idea, done!

oxtoacart commented on proxy_http.go:
```diff
@@ -11,169 +12,196 @@ import (
+)
+
+// DownstreamConn retrieves the downstream connection from the given Context.
+func DownstreamConn(ctx context.Context) net.Conn {
```
Done

oxtoacart commented on proxy_http.go:
```diff
@@ -11,169 +12,196 @@ import (
+	} else {
+		tr := &http.Transport{
+			DialContext: func(ctx context.Context, net, addr string) (net.Conn, error) {
+				return proxy.Dial(false, net, addr)
```
Good idea. I'm actually just using whatever timeout the caller passed in, even if it's bigger than 30.

oxtoacart commented on proxy_connect.go:
```diff
@@ -61,83 +36,72 @@ type connectInterceptor struct {
+		// Note - for CONNECT requests, we use the Host from the request URL, not the
+		// Host header. See discussion here:
+		// https://ask.wireshark.org/questions/22988/http-host-header-with-and-without-port-number
+		upstream, err := proxy.Dial(true, "tcp", modifiedReq.URL.Host)
```
Done.

oxtoacart commented on proxy_connect.go:
```diff
@@ -61,83 +36,72 @@ type connectInterceptor struct {
-		} else {
-			log.Error(err)
-		}
+func (proxy *proxy) dialAndCopy(addr string, downstream net.Conn) error {
```
I like the way it reads right now, leaving as is.

oxtoacart commented:
Thanks @joesis. I've addressed all comments except the keep-alive header thing, which I still need to test. There was some reason I commented it, but I can't remember why.

oxtoacart commented on proxy_http.go:
```diff
@@ -260,7 +288,7 @@ func copyHeadersForForwarding(dst, src http.Header) {
 			// the slice is short typically, don't bother sort it to speed up lookup
 			extraHopByHopHeaders = vv
-		case "Keep-Alive":
+		// case "Keep-Alive":
```
After testing, it looks like there's no reason for this to be commented out, so I reverted it to its original form.

oxtoacart commented:
@joesis Thanks again for the review, I'm done with my updates.

oxtoacart commented:
I'm going to go ahead and merge this @joesis. Thanks again for the review.
